From 3614c6c4a21208af9368fa843f0462194498bf96 Mon Sep 17 00:00:00 2001
From: ZhangPeng <zhangpeng362@huawei.com>
Date: Thu, 6 Jul 2023 20:02:55 +0800
Subject: [PATCH 631/780] [Huawei] userswap: use mmap write lock in the
 swap-out phase

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: 188960

--------------------------------

The swap-out phase does not support multithreading. We use mmap write
locks to protect the remapping of multithreaded swaps.

Signed-off-by: ZhangPeng <zhangpeng362@huawei.com>
---
 mm/userswap.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/mm/userswap.c b/mm/userswap.c
index 286260ea25be..0363b1370c9a 100644
--- a/mm/userswap.c
+++ b/mm/userswap.c
@@ -387,19 +387,19 @@ unsigned long uswap_mremap(unsigned long old_addr, unsigned long old_len,
 		return ret;
 
 	lru_add_drain_all();
-	down_read(&mm->mmap_lock);
+	down_write(&mm->mmap_lock);
 	ret = pages_can_be_swapped(mm, old_addr, len, &pages);
 	if (ret) {
-		up_read(&mm->mmap_lock);
+		up_write(&mm->mmap_lock);
 		return ret;
 	}
 
 	ret = do_user_swap(mm, old_addr, len, pages, new_addr);
-	up_read(&mm->mmap_lock);
 	/* follow_page() above increased the reference*/
 	for (i = 0; i < len / PAGE_SIZE; i++)
 		if (pages[i])
 			put_page(pages[i]);
+	up_write(&mm->mmap_lock);
 	kvfree(pages);
 	return ret;
 }
-- 
2.17.1

